File: ios/Sources/TankbookCore/Feedback/FeedbackModels.swift
import Foundation

// The feedback wire models (docs/API.md -> Feedback). Built against the
// documented contract exactly as every other client row is built against a stub
// transport: `POST /feedback` is specified and the server half is filed
// separately as PJ.20a.

/// The feedback category the user picks. The wire sends the raw string; the
/// three cases are the only values the documented contract accepts.
public enum FeedbackCategory: String, Codable, Sendable, Equatable, CaseIterable {
    case feature
    case problem
    case other
}

/// The `POST /feedback` body (docs/API.md -> Feedback). `deviceModel` rides only
/// with the user's toggle - the contract names it optional and the composer
/// attaches it only when the toggle is on. `replyTo` is an optional contact
/// address. `text` is capped at 4 KB by the composer (docs/API.md: "text ≤ 4 KB").
public struct FeedbackPayload: Codable, Sendable, Equatable, Hashable {
    public let category: FeedbackCategory
    public let text: String
    public let appVersion: String
    public let deviceModel: String?
    public let replyTo: String?

    public static let maxTextLength = 4_000

    public init(category: FeedbackCategory, text: String, appVersion: String,
                deviceModel: String?, replyTo: String?) {
        self.category = category
        self.text = String(text.prefix(Self.maxTextLength))
        self.appVersion = appVersion
        self.deviceModel = deviceModel
        self.replyTo = replyTo
    }
}

File: backend/src/Tankbook.Api/Feedback/FeedbackModels.cs
namespace Tankbook.Api.Feedback;

/// <summary>
/// The POST /feedback body (docs/API.md -> Feedback). The wire sends the
/// category as the raw string; the endpoint validates it is one of the three
/// contract values. `deviceModel` rides only with the user's toggle; `replyTo`
/// is an optional contact address. `text` is bounded by the endpoint's body
/// cap (BodySizeLimits.FeedbackBytes) - the composer caps it at 4,000
/// characters, and the client's model type documents the same contract.
/// </summary>
public sealed record FeedbackRequest(
    string? Category,
    string? Text,
    string? AppVersion,
    string? DeviceModel,
    string? ReplyTo);

/// <summary>The three categories the documented contract accepts.</summary>
public static class FeedbackCategories
{
    public const string Feature = "feature";
    public const string Problem = "problem";
    public const string Other = "other";

    public static bool IsValid(string? value)
        => value is Feature or Problem or Other;
}
